src/index.ts

```diff
@@ -178,15 +183,39 @@ export class Spectral {
     return ruleEntry.apply(opt);
   }
 
-  private _parseRuleDefinition(name: string, rule: types.Rule, format: string): IRuleEntry {
+  private _parseRuleSets(
+    rulesets: types.IRuleset[],
+    opts: { includeCurrent: boolean } = { includeCurrent: true }
+  ): { rulesets: types.IRuleset[]; functionStore: IFunctionStore; ruleStore: IRuleStore } {
+    const rSets = merge([], rulesets);
```

Comment:
Would a shallow copy work here? Merging an array feels gross.

const rSets = [...rulesets];

Reply:
Hi Marc, this actually solves a bug with mutating a deeply nested object. Imagine you pass a ruleset from outside the class and our class mutates on of the deeply nested objects. As a client of the class you may (will) not expect mutations.

Please see https://github.com/stoplightio/spectral/blob/develop/src/__tests__/spectral.test.ts#L19 if you need more details.

As a rule of thumb my opinion is that we should never mutate input parameters because it causes bugs that are hard to spot and results in an unpredictable API. 

Merge is one but not the most elegant solution of course. How about we use `immutablejs`. 
^ @marbemac @casserni?

Reply:
If we integrate something like immutable, I think that will be a larger task outside the scope of this one (may require some refactoring to using sets/gets instead of equals everywhere)